File: packages/native-helpers/swift-helper/Sources/SwiftHelper/RpcHandler.swift
import Foundation
import ObjCExceptionCatcher

/// Flexible RPC request that can parse any method string
struct FlexibleRPCRequest: Codable {
    let id: String
    let method: String
    let params: JSONAny?
}

enum RPCDispatchFailure: Swift.Error {
    case invalidParams(data: Data?, message: String)
    case internalError(data: Data?, message: String)

    var code: Int {
        switch self {
        case .invalidParams:
            return -32602
        case .internalError:
            return -32603
        }
    }

    var data: Data? {
        switch self {
        case .invalidParams(let data, _), .internalError(let data, _):
            return data
        }
    }

    var message: String {
        switch self {
        case .invalidParams(_, let message), .internalError(_, let message):
            return message
        }
    }
}

enum RPCResponseExecutor {
    static func execute(
        id: String,
        operation: () async throws -> JSONAny?
    ) async -> RPCResponseSchema {
        do {
            return RPCResponseSchema(error: nil, id: id, result: try await operation())
        } catch let failure as RPCDispatchFailure {
            let error = Error(
                code: failure.code,
                data: failure.data.flatMap {
                    try? JSONDecoder().decode(JSONAny.self, from: $0)
                },
                message: failure.message
            )
            return RPCResponseSchema(error: error, id: id, result: nil)
        } catch {
            let payload = Error(
                code: -32603,
                data: nil,
                message: "Internal error: \(error.localizedDescription)"
            )
            return RPCResponseSchema(error: payload, id: id, result: nil)
        }
    }
}

class IOBridge: NSObject {
    private let accessibilityService: AccessibilityService
    private let recordingAmbianceController: RecordingAmbianceController
    // Copy capture polls the pasteboard for up to the capture timeout; it
    // touches no AX state, so it gets its own serial queue instead of
    // blocking AccessibilityQueue (and the AX context requests behind it).
    private let copyCaptureQueue = DispatchQueue(label: "com.amical.helper.copy-capture")

    override init() {
        let accessibilityService = AccessibilityService()
        let audioService = AudioService()  // Audio preloaded here at startup
        self.accessibilityService = accessibilityService
        self.recordingAmbianceController = RecordingAmbianceController(
            playSound: { try await audioService.playSound(named: $0) },
            muteSystemAudio: { accessibilityService.muteSystemAudio() },
            restoreSystemAudio: { accessibilityService.restoreSystemAudio() },
            log: { HelperLogger.logToStderr($0) }
        )
        super.init()
    }

    private func logToStderr(_ message: String) {
        HelperLogger.logToStderr(message)
    }

    func handleRpcRequest(_ request: RPCRequestSchema) async {
        let response = await RPCResponseExecutor.execute(id: request.id) { [self] in
            try await dispatchRpcRequest(request)
        }
        sendRpcResponse(response)
    }

    private func dispatchRpcRequest(_ request: RPCRequestSchema) async throws -> JSONAny? {
        switch request.method {
        case .getAccessibilityTreeDetails:
            return try encodeResult(try await handleAccessibilityTreeDetails(request))

        case .getAccessibilityContext:
            return try encodeResult(
                try await handleGetAccessibilityContext(params: request.params))

        case .getAccessibilityStatus:
            logToStderr("[IOBridge] Handling getAccessibilityStatus for ID: \(request.id)")
            return try encodeResult(AccessibilityContextService.getAccessibilityStatus())

        case .requestAccessibilityPermission:
            logToStderr(
                "[IOBridge] Handling requestAccessibilityPermission for ID: \(request.id)")
            return try encodeResult(AccessibilityContextService.requestAccessibilityPermission())

        case .pasteText:
            return try encodeResult(try await handlePasteText(request))

        case .startRecording:
            return try encodeResult(try await handleStartRecording(request))

        case .stopRecording:
            return try encodeResult(try await handleStopRecording(request))

        case .setShortcuts:
            return try encodeResult(try handleSetShortcuts(request))

        case .setDraftEnterCapture:
            return try encodeResult(try handleSetDraftEnterCapture(request))

        case .setAllowInjectedKeys:
            return try encodeResult(try handleSetAllowInjectedKeys(request))

        case .recheckPressedKeys:
            return try encodeResult(try handleRecheckPressedKeys(request))

        case .getSelectedTextViaCopy:
            return try encodeResult(try await handleGetSelectedTextViaCopy(request))

        }
    }

    private func handleAccessibilityTreeDetails(
        _ request: RPCRequestSchema
    ) async throws -> GetAccessibilityTreeDetailsResultSchema {
        logToStderr("[IOBridge] Handling getAccessibilityTreeDetails for ID: \(request.id)")
        let params: GetAccessibilityTreeDetailsParamsSchema? = try decodeOptionalParams(
            request.params,
            method: "getAccessibilityTreeDetails"
        )

        return try await AccessibilityQueue.shared.perform { [self] in
            switch ExceptionCatcher.try({
                self.accessibilityService.fetchFullAccessibilityTree(rootId: params?.rootID)
            }) {
            case .success(let tree):
                var treeAsJSON: JSONAny?
                if let tree {
                    treeAsJSON = try self.encodeResult(tree)
                }
                return GetAccessibilityTreeDetailsResultSchema(tree: treeAsJSON)

            case .exception(let exception):
                throw RPCDispatchFailure.internalError(
                    data: self.exceptionData(exception),
                    message: "\(exception.name): \(exception.reason)"
                )
            }
        }
    }

    private func handleGetAccessibilityContext(
        params: JSONAny?
    ) async throws -> GetAccessibilityContextResult {
        logToStderr("[IOBridge] Handling getAccessibilityContext")
        let decoded: GetAccessibilityContextParams? = try decodeOptionalParams(
            params,
            method: "getAccessibilityContext"
        )
        let editableOnly = decoded?.editableOnly ?? false

        return try await AccessibilityQueue.shared.perform {
            switch ExceptionCatcher.try({
                AccessibilityContextService.getAccessibilityContext(editableOnly: editableOnly)
            }) {
            case .success(let context):
                return GetAccessibilityContextResult(context: context)

            case .exception(let exception):
                throw RPCDispatchFailure.internalError(
                    data: nil,
                    message: "\(exception.name): \(exception.reason)"
                )
            }
        }
    }

    private func handlePasteText(
        _ request: RPCRequestSchema
    ) async throws -> PasteTextResultSchema {
        logToStderr("[IOBridge] Handling pasteText for ID: \(request.id)")
        let params: PasteTextParamsSchema = try decodeRequiredParams(
            request.params,
            method: "pasteText"
        )

        return try await performOnCopyCaptureQueue { [self] in
            let preserveClipboard = params.preserveClipboard ?? true
            let success = accessibilityService.pasteText(
                transcript: params.transcript,
                preserveClipboard: preserveClipboard
            )
            return PasteTextResultSchema(
                message: success ? "Pasted successfully" : "Paste failed",
                success: success
            )
        }
    }

    private func handleStartRecording(
        _ request: RPCRequestSchema
    ) async throws -> StartRecordingResultSchema {
        logToStderr("[IOBridge] Handling startRecording for ID: \(request.id)")
        let params: StartRecordingParamsSchema = try decodeRequiredParams(
            request.params,
            method: "startRecording"
        )

        let success = await recordingAmbianceController.startRecording(
            muteSystemAudio: params.muteSystemAudio,
            muteSounds: params.muteSounds == true
        )

        return StartRecordingResultSchema(
            message: success ? "Recording started" : "Failed to mute system audio",
            success: success
        )
    }

    private func handleStopRecording(
        _ request: RPCRequestSchema
    ) async throws -> StopRecordingResultSchema {
        logToStderr("[IOBridge] Handling stopRecording for ID: \(request.id)")
        let params: StopRecordingParamsSchema = try decodeRequiredParams(
            request.params,
            method: "stopRecording"
        )

        let success = await recordingAmbianceController.stopRecording(
            wasMuted: params.wasMuted,
            muteSounds: params.muteSounds == true
        )

        return StopRecordingResultSchema(
            message: success ? "Recording stopped" : "Failed to restore system audio",
            success: success
        )
    }

    private func handleSetShortcuts(
        _ request: RPCRequestSchema
    ) throws -> SetShortcutsResultSchema {
        logToStderr("[IOBridge] Handling setShortcuts for ID: \(request.id)")
        let params: SetShortcutsParamsSchema = try decodeRequiredParams(
            request.params,
            method: "setShortcuts"
        )
        ShortcutManager.shared.setShortcuts(
            subsetChords: params.subsetChords,
            exactChords: params.exactChords
        )
        return SetShortcutsResultSchema(success: true)
    }

    private func handleSetDraftEnterCapture(
        _ request: RPCRequestSchema
    ) throws -> SetDraftEnterCaptureResultSchema {
        logToStderr("[IOBridge] Handling setDraftEnterCapture for ID: \(request.id)")
        let params: SetDraftEnterCaptureParamsSchema = try decodeRequiredParams(
            request.params,
            method: "setDraftEnterCapture"
        )
        ShortcutManager.shared.setDraftEnterCapture(params.enabled)
        return SetDraftEnterCaptureResultSchema(success: true)
    }

    private func handleSetAllowInjectedKeys(
        _ request: RPCRequestSchema
    ) throws -> SetAllowInjectedKeysResultSchema {
        logToStderr(
            "[IOBridge] Handling setAllowInjectedKeys (no-op on macOS) for ID: \(request.id)")
        let _: SetAllowInjectedKeysParamsSchema = try decodeRequiredParams(
            request.params,
            method: "setAllowInjectedKeys"
        )
        return SetAllowInjectedKeysResultSchema(success: true)
    }

    private func handleRecheckPressedKeys(
        _ request: RPCRequestSchema
    ) throws -> RecheckPressedKeysResultSchema {
        logToStderr("[IOBridge] Handling recheckPressedKeys for ID: \(request.id)")
        let params: RecheckPressedKeysParamsSchema = try decodeRequiredParams(
            request.params,
            method: "recheckPressedKeys"
        )
        let staleKeyCodes = ShortcutManager.shared.getStalePressedKeyCodes(
            params.pressedKeyCodes
        )
        return RecheckPressedKeysResultSchema(staleKeyCodes: staleKeyCodes)
    }

    private func handleGetSelectedTextViaCopy(
        _ request: RPCRequestSchema
    ) async throws -> GetSelectedTextViaCopyResultSchema {
        logToStderr("[IOBridge] Handling getSelectedTextViaCopy for ID: \(request.id)")
        return try await performOnCopyCaptureQueue { [self] in
            accessibilityService.getSelectedTextViaCopy()
        }
    }

    private func decodeRequiredParams<T: Decodable>(
        _ params: JSONAny?,
        method: String
    ) throws -> T {
        guard let params else {
            throw RPCDispatchFailure.invalidParams(
                data: nil,
                message: "Missing params for \(method)"
            )
        }
        var payload: Data?
        do {
            let data = try JSONEncoder().encode(params)
            payload = data
            return try JSONDecoder().decode(T.self, from: data)
        } catch {
            throw RPCDispatchFailure.invalidParams(
                data: payload,
                message: "Invalid params for \(method): \(error.localizedDescription)"
            )
        }
    }

    private func decodeOptionalParams<T: Decodable>(
        _ params: JSONAny?,
        method: String
    ) throws -> T? {
        guard let params else { return nil }
        var payload: Data?
        do {
            let data = try JSONEncoder().encode(params)
            payload = data
            return try JSONDecoder().decode(T.self, from: data)
        } catch {
            throw RPCDispatchFailure.invalidParams(
                data: payload,
                message: "Invalid params for \(method): \(error.localizedDescription)"
            )
        }
    }

    private func encodeResult<T: Encodable>(_ result: T) throws -> JSONAny {
        do {
            let data = try JSONEncoder().encode(result)
            return try JSONDecoder().decode(JSONAny.self, from: data)
        } catch {
            throw RPCDispatchFailure.internalError(
                data: nil,
                message: "Error encoding result: \(error.localizedDescription)"
            )
        }
    }

    private func performOnCopyCaptureQueue<T>(
        _ operation: @escaping () throws -> T
    ) async throws -> T {
        try await withCheckedThrowingContinuation { continuation in
            copyCaptureQueue.async {
                do {
                    continuation.resume(returning: try operation())
                } catch {
                    continuation.resume(throwing: error)
                }
            }
        }
    }

    private func exceptionData(_ exception: CaughtException) -> Data? {
        let payload: [String: Any] = [
            "name": exception.name,
            "reason": exception.reason,
            "callStack": exception.callStack.prefix(10).joined(separator: "\n"),
        ]
        return try? JSONSerialization.data(withJSONObject: payload)
    }

    private func sendRpcResponse(_ response: RPCResponseSchema) {
        do {
            let responseData = try JSONEncoder().encode(response)
            if let responseString = String(data: responseData, encoding: .utf8) {
                logToStderr("[Swift Biz Logic] FINAL JSON RESPONSE to stdout: \(responseString)")
                StdoutWriter.writeLine(responseString)
            }
        } catch {
            logToStderr("Error encoding RpcResponse: \(error.localizedDescription)")
        }
    }

    // Main loop for processing RPC requests from stdin
    func processRpcRequests() {
        logToStderr("IOBridge: Starting RPC request processing loop.")
        while let line = readLine(strippingNewline: true) {
            guard !line.isEmpty, let data = line.data(using: .utf8) else {
                logToStderr("Warning: Received empty or non-UTF8 line on stdin.")
                continue
            }

            do {
                let request = try JSONDecoder().decode(RPCRequestSchema.self, from: data)
                logToStderr(
                    "IOBridge: Received RPC Request ID \(request.id), Method: \(request.method)"
                )
                Task(priority: .high) { [self] in
                    await handleRpcRequest(request)
                }
            } catch {
                logToStderr(
                    "Error decoding RpcRequest from stdin: \(error.localizedDescription). Line: \(line)"
                )
            }
        }
        logToStderr("IOBridge: RPC request processing loop finished (stdin closed).")
    }
}

File: packages/native-helpers/windows-helper/src/RpcHandler.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;
using WindowsHelper.Models;
using WindowsHelper.Services;

namespace WindowsHelper
{
    internal sealed class RpcDispatchException : Exception
    {
        internal long RpcCode { get; }
        internal object? RpcData { get; }

        internal RpcDispatchException(long code, string message, object? data = null)
            : base(message)
        {
            RpcCode = code;
            RpcData = data;
        }
    }

    internal static class RpcResponseExecutor
    {
        internal static async Task<string> ExecuteAsync(
            string id,
            Func<Task<object?>> operation,
            JsonSerializerOptions jsonOptions)
        {
            RpcResponse response;
            try
            {
                response = new RpcResponse
                {
                    Id = id,
                    Result = await operation()
                };
            }
            catch (RpcDispatchException ex)
            {
                response = new RpcResponse
                {
                    Id = id,
                    Error = new Error
                    {
                        Code = ex.RpcCode,
                        Message = ex.Message,
                        Data = ex.RpcData
                    }
                };
            }
            catch (Exception ex)
            {
                response = InternalError(id, ex);
            }

            try
            {
                return JsonSerializer.Serialize(response, jsonOptions);
            }
            catch (Exception ex)
            {
                return JsonSerializer.Serialize(InternalError(id, ex), jsonOptions);
            }
        }

        private static RpcResponse InternalError(string id, Exception exception)
        {
            return new RpcResponse
            {
                Id = id,
                Error = new Error
                {
                    Code = -32603,
                    Message = $"Internal error: {exception.Message}"
                }
            };
        }
    }

    public class RpcHandler : IDisposable
    {
        private readonly JsonSerializerOptions jsonOptions;
        private readonly AccessibilityService accessibilityService;
        private readonly AudioService audioService;
        private readonly StaThreadRunner? staRunner;
        private bool disposed;

        public RpcHandler(StaThreadRunner? staRunner, ClipboardService clipboardService)
        {
            this.staRunner = staRunner;
            jsonOptions = WindowsHelper.Models.Converter.Settings;
            accessibilityService = new AccessibilityService(clipboardService);
            audioService = new AudioService();

            if (staRunner != null)
            {
                LogToStderr("RpcHandler: STA thread dispatch enabled via StaThreadRunner");
            }
        }

        public void Dispose()
        {
            if (disposed) return;
            disposed = true;
        }

        public void ProcessRpcRequests(CancellationToken cancellationToken)
        {
            LogToStderr("RpcHandler: Starting RPC request processing loop.");

            try
            {
                string? line;
                while (!cancellationToken.IsCancellationRequested && (line = Console.ReadLine()) != null)
                {
                    if (string.IsNullOrWhiteSpace(line))
                    {
                        LogToStderr("Warning: Received empty line on stdin.");
                        continue;
                    }

                    try
                    {
                        var request = JsonSerializer.Deserialize<RpcRequest>(line, jsonOptions);
                        if (request == null)
                        {
                            LogToStderr($"Error decoding RpcRequest from stdin: decoded request was null. Line: {line}");
                            continue;
                        }

                        LogToStderr($"RpcHandler: Received RPC Request ID {request.Id}, Method: {request.Method}");
                        _ = Task.Run(
                            () => ProcessAndRespondAsync(request),
                            cancellationToken
                        );
                    }
                    catch (Exception ex)
                    {
                        // Generated converters can throw exceptions other than JsonException.
                        // Keep every decode failure inside this iteration so the reader survives.
                        LogToStderr($"Error decoding RpcRequest from stdin: {ex.Message}. Line: {line}");
                    }
                }
            }
            catch (Exception ex)
            {
                LogToStderr($"Fatal error reading RPC stdin: {ex.Message}");
            }

            LogToStderr("RpcHandler: RPC request processing loop finished.");
        }

        private async Task ProcessAndRespondAsync(RpcRequest request)
        {
            var responseJson = await RpcResponseExecutor.ExecuteAsync(
                request.Id.ToString(),
                () => DispatchRpcRequestAsync(request),
                jsonOptions
            );
            SendRpcResponse(responseJson);
        }

        private async Task<object?> DispatchRpcRequestAsync(RpcRequest request)
        {
            switch (request.Method)
            {
                case Method.GetAccessibilityTreeDetails:
                    return await HandleGetAccessibilityTreeDetails(request);

                case Method.GetAccessibilityContext:
                    return await HandleGetAccessibilityContext(request);

                case Method.PasteText:
                    return HandlePasteText(request);

                case Method.StartRecording:
                    return await HandleStartRecording(request);

                case Method.StopRecording:
                    return HandleStopRecording(request);

                case Method.SetShortcuts:
                    return HandleSetShortcuts(request);

                case Method.SetDraftEnterCapture:
                    return HandleSetDraftEnterCapture(request);

                case Method.SetAllowInjectedKeys:
                    return HandleSetAllowInjectedKeys(request);

                case Method.RecheckPressedKeys:
                    return HandleRecheckPressedKeys(request);

                case Method.GetSelectedTextViaCopy:
                    return HandleGetSelectedTextViaCopy(request);

                case Method.GetAccessibilityStatus:
                case Method.RequestAccessibilityPermission:
                    throw new RpcDispatchException(
                        -32601,
                        $"Method not found: {request.Method}"
                    );

                default:
                    throw new RpcDispatchException(
                        -32601,
                        $"Method not found: {request.Method}"
                    );
            }
        }

        private async Task<object?> HandleGetAccessibilityTreeDetails(RpcRequest request)
        {
            LogToStderr($"Handling getAccessibilityTreeDetails for ID: {request.Id}");
            var parameters = DecodeOptionalParams<GetAccessibilityTreeDetailsParams>(
                request,
                "getAccessibilityTreeDetails"
            );
            var tree = await Task.Run(
                () => accessibilityService.FetchAccessibilityTree(parameters?.RootId)
            );
            return new GetAccessibilityTreeDetailsResult { Tree = tree };
        }

        private async Task<object?> HandleGetAccessibilityContext(RpcRequest request)
        {
            LogToStderr($"Handling getAccessibilityContext for ID: {request.Id}");
            var parameters = DecodeOptionalParams<GetAccessibilityContextParams>(
                request,
                "getAccessibilityContext"
            );
            var context = await Task.Run(
                () => accessibilityService.GetAccessibilityContext(parameters?.EditableOnly ?? false)
            );
            return new GetAccessibilityContextResult { Context = context };
        }

        private object HandleGetSelectedTextViaCopy(RpcRequest request)
        {
            LogToStderr($"Handling getSelectedTextViaCopy for ID: {request.Id}");
            return accessibilityService.GetSelectedTextViaCopy();
        }

        private object HandlePasteText(RpcRequest request)
        {
            LogToStderr($"Handling pasteText for ID: {request.Id}");
            var parameters = DecodeRequiredParams<PasteTextParams>(
                request,
                "pasteText",
                "transcript"
            );
            var preserveClipboard = parameters.PreserveClipboard ?? true;
            var success = accessibilityService.PasteText(
                parameters.Transcript,
                preserveClipboard,
                out var errorMessage
            );
            return new PasteTextResult
            {
                Success = success,
                Message = success
                    ? (errorMessage ?? "Pasted successfully")
                    : (errorMessage ?? "Paste failed")
            };
        }

        private async Task<object?> HandleStartRecording(RpcRequest request)
        {
            LogToStderr($"Handling startRecording for ID: {request.Id}");
            var parameters = DecodeRequiredParams<StartRecordingParams>(
                request,
                "startRecording",
                "muteSystemAudio"
            );

            if (parameters.MuteSounds != true)
            {
                await audioService.PlaySound("rec-start");
            }

            var success = true;
            if (parameters.MuteSystemAudio)
            {
                success = audioService.MuteSystemAudio();
            }

            return new StartRecordingResult
            {
                Success = success,
                Message = success ? "Recording started" : "Failed to mute system audio"
            };
        }

        private object HandleStopRecording(RpcRequest request)
        {
            LogToStderr($"Handling stopRecording for ID: {request.Id}");
            var parameters = DecodeRequiredParams<StopRecordingParams>(
                request,
                "stopRecording",
                "wasMuted"
            );

            var success = true;
            if (parameters.WasMuted)
            {
                success = audioService.RestoreSystemAudio();
            }

            if (parameters.MuteSounds != true)
            {
                _ = PlayStopSoundAsync();
            }

            return new StopRecordingResult
            {
                Success = success,
                Message = success ? "Recording stopped" : "Failed to restore system audio"
            };
        }

        private async Task PlayStopSoundAsync()
        {
            try
            {
                await audioService.PlaySound("rec-stop");
            }
            catch (Exception ex)
            {
                LogToStderr($"Error playing rec-stop: {ex.Message}");
            }
        }

        private object HandleSetShortcuts(RpcRequest request)
        {
            LogToStderr($"[RpcHandler] Handling setShortcuts for ID: {request.Id}");
            var parameters = DecodeRequiredParams<SetShortcutsParams>(
                request,
                "setShortcuts",
                "subsetChords",
                "exactChords"
            );
            ShortcutManager.Instance.SetShortcuts(
                ConvertChords(parameters.SubsetChords),
                ConvertChords(parameters.ExactChords)
            );
            return new SetShortcutsResult { Success = true };
        }

        private object HandleSetDraftEnterCapture(RpcRequest request)
        {
            LogToStderr($"[RpcHandler] Handling setDraftEnterCapture for ID: {request.Id}");
            var parameters = DecodeRequiredParams<SetDraftEnterCaptureParams>(
                request,
                "setDraftEnterCapture",
                "enabled"
            );
            ShortcutManager.Instance.SetDraftEnterCapture(parameters.Enabled);
            return new SetDraftEnterCaptureResult { Success = true };
        }

        private object HandleSetAllowInjectedKeys(RpcRequest request)
        {
            LogToStderr($"[RpcHandler] Handling setAllowInjectedKeys for ID: {request.Id}");
            var parameters = DecodeRequiredParams<SetAllowInjectedKeysParams>(
                request,
                "setAllowInjectedKeys",
                "enabled"
            );
            ShortcutManager.Instance.SetAllowInjectedKeys(parameters.Enabled);
            return new SetAllowInjectedKeysResult { Success = true };
        }

        private object HandleRecheckPressedKeys(RpcRequest request)
        {
            LogToStderr($"Handling recheckPressedKeys for ID: {request.Id}");
            var parameters = DecodeRequiredParams<RecheckPressedKeysParams>(
                request,
                "recheckPressedKeys",
                "pressedKeyCodes"
            );
            var staleKeyCodes = ShortcutManager.Instance.GetStalePressedKeyCodes(
                (parameters.PressedKeyCodes ?? new List<long>()).Select(keyCode => (int)keyCode)
            );
            return new RecheckPressedKeysResult
            {
                StaleKeyCodes = staleKeyCodes.Select(keyCode => (long)keyCode).ToList()
            };
        }

        private T DecodeRequiredParams<T>(
            RpcRequest request,
            string method,
            params string[] requiredProperties)
        {
            if (request.Params == null)
            {
                throw InvalidParams(request, $"Missing params for {method}");
            }

            try
            {
                var json = JsonSerializer.Serialize(request.Params, jsonOptions);
                using var document = JsonDocument.Parse(json);
                if (document.RootElement.ValueKind != JsonValueKind.Object)
                {
                    throw InvalidParams(request, $"Invalid params for {method}: expected an object");
                }

                foreach (var propertyName in requiredProperties)
                {
                    if (!document.RootElement.TryGetProperty(propertyName, out var property)
                        || property.ValueKind == JsonValueKind.Null)
                    {
                        throw InvalidParams(
                            request,
                            $"Invalid params for {method}: missing {propertyName}"
                        );
                    }
                }

                return JsonSerializer.Deserialize<T>(json, jsonOptions)
                    ?? throw InvalidParams(request, $"Invalid params for {method}");
            }
            catch (RpcDispatchException)
            {
                throw;
            }
            catch (Exception ex)
            {
                throw InvalidParams(request, $"Invalid params for {method}: {ex.Message}");
            }
        }

        private T? DecodeOptionalParams<T>(RpcRequest request, string method)
        {
            if (request.Params == null) return default;

            try
            {
                var json = JsonSerializer.Serialize(request.Params, jsonOptions);
                return JsonSerializer.Deserialize<T>(json, jsonOptions);
            }
            catch (Exception ex)
            {
                throw InvalidParams(request, $"Invalid params for {method}: {ex.Message}");
            }
        }

        private static RpcDispatchException InvalidParams(RpcRequest request, string message)
        {
            return new RpcDispatchException(-32602, message, request.Params);
        }

        private void SendRpcResponse(string responseJson)
        {
            try
            {
                LogToStderr($"[RpcHandler] Sending response to stdout: {responseJson}");
                StdoutWriter.WriteLine(responseJson);
            }
            catch (Exception ex)
            {
                LogToStderr($"Error writing RpcResponse: {ex.Message}");
            }
        }

        private void LogToStderr(string message)
        {
            HelperLogger.LogToStderr(message);
        }

        private static int[] ConvertKeycodes(List<long>? keycodes)
        {
            if (keycodes == null || keycodes.Count == 0) return Array.Empty<int>();
            return keycodes.Select(keycode => (int)keycode).ToArray();
        }

        private static int[][] ConvertChords(List<List<long>>? chords)
        {
            if (chords == null) return Array.Empty<int[]>();
            return chords.Select(ConvertKeycodes).ToArray();
        }
    }
}
